### PR TITLE
RAC4737: Add Unit-Test step in Nightly Build & Sprint Regression

### PR DESCRIPTION
**Pull request by PengTian0**
[RAC4737](https://rackhd.atlassian.net/browse/RAC-4737): Add Unit-Test step in Nightly Build & Sprint Regression

**Test Done**:
SprintTag, SprintRelease and MasterCI passed.

### CODE REVIEW

**Comment by PengTian0** [comment]
@panpan0000 @changev @keedya @anhou @yyscamper

File: jobs/ShareMethod.groovy [hunk 1]
```diff
@@ -39,6 +39,23 @@ def getLockedResourceName(resources,label_name){
...
+    resources = getLockedResourceName(lock_resources,label_name)
+    def available_resources = resources - used_resources
+    if(available_resources.size > 0){
+        used_resources.add(available_resources[0])
+        resource_name = available_resources[0]
+        return resource_name
+    }
+    else{
+        sh '''
+        echo "There is no available resources for'''+"$label_name"+'''
```
Comment (yyscamper) [inline jobs/ShareMethod.groovy]: Seems there is mismatched `'''` and `"`, at least for the `"` after `echo`

File: jobs/UnitTest/UnitTest.groovy [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                    --test-result-file xunit-reports/'''+"${repo_name}"+'''.xml  \
+                    --parameters-file downstream_file
+                    '''
+                    int failure_count = 0
+                    if(fileExists ("downstream_file")) {
+                        def props = readProperties file: "downstream_file"
+                        failure_count = "${props.failures}".toInteger()
+                    }
+                    if (failure_count > 0){
+                        currentBuild.result = "SUCCESS"
```
Comment (yyscamper) [inline jobs/UnitTest/UnitTest.groovy]: there is failure but why the result is success? typo?

File: jobs/pr_gate/unit_test.groovy [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+            def repo_dir = pwd()
+            def prgate_unit_test = load("jobs/UnitTest/PRGateUnitTest.groovy")
+            if("${stash_manifest_name}" != null && "${stash_manifest_name}" != "null"){
+                if("${stash_manifest_path}" != null && "${stash_manifest_path}" != "null"){
+                    prgate_unit_test.runTest("${stash_manifest_name}", "${stash_manifest_path}", repo_dir)
+                }
+                else{
+                    sh '''
+                    echo "Please provider argument stash_manifest_path if you provided argument stash_manifest_name"
```
Comment (yyscamper) [inline jobs/pr_gate/unit_test.groovy]: typo provider => provide

File: jobs/UnitTest/UnitTest.groovy [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                    junit 'xunit-reports/'+"${repo_name}.xml"
+
+                    sh '''
+                    ./build-config/build-release-tools/application/parse_test_results.py \
+                    --test-result-file xunit-reports/'''+"${repo_name}"+'''.xml  \
+                    --parameters-file downstream_file
+                    '''
+                    int failure_count = 0
+                    if(fileExists ("downstream_file")) {
+                        def props = readProperties file: "downstream_file"
```
Comment (yyscamper) [inline jobs/UnitTest/UnitTest.groovy]: `readProperties` is a function? if so, use `()` to enclose the argument will be better.

File: jobs/ShareMethod.groovy (hunk 1 above)
Comment (PengTian0) [reply jobs/ShareMethod.groovy]: Thanks, Fixed

File: jobs/pr_gate/unit_test.groovy (hunk 3 above)
Comment (PengTian0) [reply jobs/pr_gate/unit_test.groovy]: Thanks, fixed

File: jobs/UnitTest/UnitTest.groovy (hunk 2 above)
Comment (PengTian0) [reply jobs/UnitTest/UnitTest.groovy]: Thanks!
Fixed.

File: jobs/UnitTest/UnitTest.groovy (hunk 4 above)
Comment (PengTian0) [reply jobs/UnitTest/UnitTest.groovy]: ```readProperties``` is a step of Jenkins Pipeline.
The grammar of the step doesn't contain ```()```

**Review by yyscamper — APPROVED** [review]

**Comment by PengTian0** [comment]
test this please

**Comment by PengTian0** [comment]
test this please

**Comment by PengTian0** [comment]
test this please

**Comment by PengTian0** [comment]
test this please

File: jobs/UnitTest/PRGateUnitTest.groovy [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+def runTest(String manifest_name, String manifest_path, String repo_dir){
+    setRepoDir(repo_dir)
+    setUnitTest()
+    setManifest(manifest_name, manifest_path)
+    setTestRepos()
+    unit_test.runTest(manifest_name, manifest_path, repo_dir)
+}
+
+return this
```
Comment (panpan0000) [inline jobs/UnitTest/PRGateUnitTest.groovy:59]: Niubility!

File: jobs/UnitTest/PRGateUnitTest.groovy [hunk 6]
```diff
@@ -0,0 +1,51 @@
+String stash_manifest_name
```
Comment (panpan0000) [inline jobs/UnitTest/PRGateUnitTest.groovy:3]: Can you add comments either in PR or code to describe the class/groovy releationship?

File: jobs/UnitTest/PRGateUnitTest.groovy [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    }
+    env.MANIFEST_FILE_PATH = repo_dir + "/${this.stash_manifest_path}"
+    sh '''#!/bin/bash
+    pushd ''' + "$repo_dir" + '''
+    ./build-release-tools/HWIMO-BUILD ./build-release-tools/application/parse_manifest.py \
+    --manifest-file $MANIFEST_FILE_PATH \
+    --parameters-file downstream_file
+    '''
+    def repos_need_unit_test = ""
+    if(fileExists ('downstream_file')) {
```
Comment (panpan0000) [inline jobs/UnitTest/PRGateUnitTest.groovy:40]: can you please explain the purpose and magic of downstream_file ? (comment in code)

**Review by panpan0000 — APPROVED** [review]

File: jobs/UnitTest/PRGateUnitTest.groovy (hunk 7 above)
Comment (PengTian0) [reply jobs/UnitTest/PRGateUnitTest.groovy:40]: Sure~

File: jobs/UnitTest/PRGateUnitTest.groovy (hunk 6 above)
Comment (PengTian0) [reply jobs/UnitTest/PRGateUnitTest.groovy:3]: Thanks, Fixed.

**Pull request merged**